backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarWriteFilesExec.scala

```diff
@@ -0,0 +1,151 @@
+        // The fragments column contains detailed information about the file writes.
+        // The json can be as following:
+        // {
+        //     "inMemoryDataSizeInBytes":0,
+        //     "containsNumberedFileNames":true,
+        //     "onDiskDataSizeInBytes":307,
+        //     "fileWriteInfos":[
+        //         {
+        //             "fileSize":307,
+        //             "writeFileName":
+        //                "Gluten_Stage_1_TID_2_0_2_d1db3b31-4f99-41cb-a4e7-3b8607506168.parquet",
+        //             "targetFileName":
+        //                "Gluten_Stage_1_TID_2_0_2_d1db3b31-4f99-41cb-a4e7-3b8607506168.parquet"
+        //         }
+        //     ],
+        //     "writePath":"file:/home/gluten/spark-warehouse/inserttable/part1=1/part2=1",
+        //     "rowCount":1,
+        //     "targetPath":"file:/home/gluten/spark-warehouse/inserttable/part1=1/part2=1",
+        //     "updateMode":"NEW",
+        //     "name":"part1=1/part2=1"
+        // }
+        if (iter.hasNext) {
+          val cb = iter.next()
+          val loadedCb = ColumnarBatches.ensureLoaded(ArrowBufferAllocators.contextInstance, cb)
+
+          val numRows = loadedCb.column(0).getLong(0)
+
+          var updatedPartitions = Set.empty[String]
+          val addedAbsPathFiles: mutable.Map[String, String] = mutable.Map[String, String]()
+          var numBytes = 0L
+          for (i <- 0 until loadedCb.numRows() - 1) {
+            val fragments = loadedCb.column(1).getUTF8String(i + 1)
+            val objectMapper = new ObjectMapper()
+            val jsonObject = objectMapper.readTree(fragments.toString)
+
+            val fileWriteInfos = jsonObject.get("fileWriteInfos").elements()
+            if (jsonObject.get("fileWriteInfos").elements().hasNext) {
+              val writeInfo = fileWriteInfos.next();
+              numBytes += writeInfo.get("fileSize").size()
+              // Get partition information.
+              if (jsonObject.get("name").textValue().nonEmpty) {
+                val targetFileName = writeInfo.get("targetFileName").textValue()
+                val partitionDir = jsonObject.get("name").textValue()
+                updatedPartitions += partitionDir
+                val tmpOutputPath =
+                  writeFilesSpec.description.path + "/" + partitionDir + "/" + targetFileName
+                val absOutputPathObject =
+                  writeFilesSpec.description.customPartitionLocations.get(
+                    PartitioningUtils.parsePathFragment(partitionDir))
+                if (absOutputPathObject.nonEmpty) {
+                  val absOutputPath = absOutputPathObject.get + "/" + targetFileName
+                  addedAbsPathFiles(tmpOutputPath) = absOutputPath
+                }
+              }
+            }
+          }
+
+          // TODO: need to get the partition Internal row?
+          val stats = BasicWriteTaskStats(Seq.empty, (numRows - 1).toInt, numBytes, numRows)
```

Comment:
why does `(numRows - 1).toInt` mean `numFiles` ?

Reply:
@ulysses-you The `fragments ` size is the `numFiles`. Velox set the `numOutputRows ` as the `fragements .size +1`  [here](https://github.com/facebookincubator/velox/blob/main/velox/exec/TableWriter.cpp#L183). So we can get the `numFiles `as `numRows -1`.